File: python/problems/leetcode/70-Climbing-Stairs.py
# Time: O(N), Space: O(N) because we create a cache
class Solution:
    def climbStairs(self, n: int) -> int:
        def memo(n, cache):
            if n <= 2:
                return n
            if n in cache:
                return cache[n]
            cache[n] = memo(n - 1, cache) + memo(n - 2, cache)
            return cache[n]

        return memo(n, {})


# Time: O(N), Space: O(1) because we create a array of size 2 and we only track those 2 values
class Solution:
    def climbStairs(self, n: int) -> int:
        if n < 1:
            return 1
        dp = [1] * 2
        for i in range(2, n + 1):
            dp[1], dp[0] = dp[1] + dp[0], dp[1]
        return dp[1]

File: cpp/problems/leetcode/70-Climbing-Stairs.cpp
#include <bits/stdc++.h>
// Time: O(N), Space: O(N) bc we create a cache
class Solution {
public:
  int memo(int n, map<int, int> &cache) {
    if (n <= 2) {
      return n;
    } else if (cache.find(n) != cache.end()) {
      return cache[n];
    }
    cache[n] = memo(n - 1, cache) + memo(n - 2, cache);
    return cache[n];
  }

  int climbStairs(int n) {
    map<int, int> cache;
    return memo(n, cache);
  }
};

// Time: O(N), Space: O(1), no arrays
class Solution {
public:
  int climbStairs(int n) {
    if (n == 1) {
      return 1;
    }
    if (n == 2) {
      return 2;
    }

    int first = 1;
    int second = 2;

    int result = 0;

    for (int i = 2; i < n; i++) {
      result = first + second;
      first = second;
      second = result;
    }

    return result;
  }
};